src/storage/txn/commands/flashback_to_version_read_phase.rs

```diff
@@ -58,62 +91,101 @@ impl<S: Snapshot> ReadCommand<S> for FlashbackToVersionReadPhase {
-            statistics,
-        )?;
-        // Scan the writes.
-        let (mut key_old_writes, has_remain_writes) = flashback_to_version_read_write(
-            &mut reader,
-            key_locks.len(),
-            &self.next_write_key,
-            &self.end_key,
-            self.version,
-            self.start_ts,
-            self.commit_ts,
-            statistics,
-        )?;
-        tls_collect_keyread_histogram_vec(
-            self.tag().get_str(),
-            (key_locks.len() + key_old_writes.len()) as f64,
-        );
-
-        if key_locks.is_empty() && key_old_writes.is_empty() {
-            Ok(ProcessResult::Res)
-        } else {
-            let next_lock_key = if has_remain_locks {
-                key_locks.last().map(|(key, _)| key.clone())
-            } else {
-                None
-            };
-            let next_write_key = if has_remain_writes && !key_old_writes.is_empty() {
-                key_old_writes.pop().map(|(key, _)| key)
-            } else if has_remain_writes && key_old_writes.is_empty() {
-                // We haven't read any write yet, so we need to read the writes in the next
-                // batch later.
-                self.next_write_key
+        // Separate the lock and write flashback to prevent from putting two writes for
+        // the same key in a single batch to make the TiCDC panic.
+        let next_state = match self.state {
+            FlashbackToVersionState::ScanLock { next_lock_key, .. } => {
+                if next_lock_key.is_none() {
+                    read_again = true;
+                    FlashbackToVersionState::ScanWrite {
+                        next_write_key: Some(self.start_key.clone()),
+                        key_old_writes: Vec::new(),
+                    }
+                } else {
+                    let (mut key_locks, has_remain_locks) = flashback_to_version_read_lock(
+                        &mut reader,
+                        &next_lock_key.unwrap(),
+                        &self.end_key,
+                        statistics,
+                    )?;
+                    if key_locks.is_empty() && !has_remain_locks {
+                        // No more locks to flashback, continue to scan the writes.
+                        read_again = true;
+                        FlashbackToVersionState::ScanWrite {
+                            next_write_key: Some(self.start_key.clone()),
+                            key_old_writes: Vec::new(),
+                        }
+                    } else {
+                        tls_collect_keyread_histogram_vec(tag, key_locks.len() as f64);
+                        FlashbackToVersionState::ScanLock {
+                            next_lock_key: if has_remain_locks {
+                                // The batch must be full.
+                                assert_eq!(key_locks.len(), FLASHBACK_BATCH_SIZE);
+                                key_locks.pop().map(|(key, _)| key)
```

Comment:
If it has remain locks, `key_locks` must be not empty. `next_write_key` and `next_lock_key` don't have to be `Option`

Reply:
Very nice catch 👍